dsfs/dataset.go

```diff
@@ -310,33 +247,96 @@ func prepareDataset(store cafs.Filestore, ds *dataset.Dataset, df cafs.File, pri
 		}
 	}
 
-	data, err := ioutil.ReadAll(df)
+	errR, errW := io.Pipe()
+	entryR, entryW := io.Pipe()
+	hashR, hashW := io.Pipe()
+	done := make(chan error)
+
+	go setErrCount(ds, cafs.NewMemfileReader(df.FileName(), errR), mu, done)
+	go setEntryCount(ds, cafs.NewMemfileReader(df.FileName(), entryR), mu, done)
+	go setChecksumAndStats(ds, cafs.NewMemfileReader(df.FileName(), hashR), &buf, mu, done)
+
+	go func() {
+		// pipes must be manually closed to trigger EOF
+		defer errW.Close()
+		defer entryW.Close()
+		defer hashW.Close()
+
+		// allocate a multiwriter that writes to each pipe when
+		// mw.Write() is called
+		mw := io.MultiWriter(errW, entryW, hashW)
+		// copy file bytes to multiwriter from input file
+		io.Copy(mw, df)
+	}()
+
+	for i := 3; i > 0; i-- {
+		if err := <-done; err != nil {
+			return nil, "", err
+		}
+	}
+
+	//get auto commit message if necessary
+	diffDescription, err := generateCommitMsg(store, ds)
+	if err != nil {
+		log.Debug(err.Error())
+		return nil, "", fmt.Errorf("%s", err.Error())
+	}
+	if diffDescription == "" {
+		return nil, "", fmt.Errorf("error saving: no changes detected")
+	}
+
+	if ds.Commit.Title == "" && ds.Commit.Message != "" {
```

Comment:
This condition is already being handled inside cleanTitleAndMessage. How about, remove this and the next condition, pass diffDescription as a third parameter to cleanTitleAndMessage, using it only if both Title and Message are blank.